contracts/src/lib/LibGateway.sol

```diff
@@ -386,20 +400,149 @@ library LibGateway {
+     *
+     * @dev It also validates that destination subnet ID is not empty
+     *      and not equal to the current network.
+     *      This function assumes that the funds inside `value` have been
+     *      conveniently minted or burnt already and the message is free to
+     *      use them (see execBottomUpMsgBatch for reference).
+     *  @param crossMessage The cross-network message to commit.
+     *  @return shouldBurn A Boolean that indicates if the input amount should be burned.
+     */
+    function commitCrossMessage(IpcEnvelope memory crossMessage) internal returns (bool shouldBurn) {
+        GatewayActorStorage storage s = LibGatewayActorStorage.appStorage();
+        SubnetID memory to = crossMessage.to.subnetId;
+        if (to.isEmpty()) {
+            revert InvalidCrossMsgDstSubnet();
+        }
+        // destination is the current network, you are better off with a good old message, no cross needed
+        if (to.equals(s.networkName)) {
+            revert CannotSendCrossMsgToItself();
+        }
+
+        SubnetID memory from = crossMessage.from.subnetId;
+        IPCMsgType applyType = crossMessage.applyType(s.networkName);
+
+        // Are we the LCA? (Lowest Common Ancestor)
+        bool isLCA = to.commonParent(from).equals(s.networkName);
+
+        // Even if multi-level messaging is enabled, we reject the xnet message
+        // as soon as we learn that one of the networks involved use an ERC20 supply source.
+        // This will block propagation on the first step, or the last step.
+        //
+        // TODO IPC does not implement fault handling yet, so if the message fails
+        //  to propagate, the user won't be able to reclaim funds. That's one of the
+        //  reasons xnet messages are disabled by default.
+
+        bool reject = false;
+        if (applyType == IPCMsgType.BottomUp) {
+            // We're traversing up, so if we're the first hop, we reject if the subnet was ERC20.
+            // If we're not the first hop, a child propagated this to us, they made a mistake and
+            // and we don't have enough info to evaluate.
+            reject = from.getParentSubnet().equals(s.networkName) && from.getActor().hasSupplyOfKind(SupplyKind.ERC20);
+        } else if (applyType == IPCMsgType.TopDown) {
+            // We're traversing down.
+            // Check the next subnet (which can may be the destination subnet).
+            reject = to.down(s.networkName).getActor().hasSupplyOfKind(SupplyKind.ERC20);
+        }
+        if (reject && crossMessage.kind == IpcMsgKind.Transfer) {
+            revert MethodNotAllowed("propagation of `Transfer` messages not suppported for subnets with ERC20 supply");
+        }
+
+        // If the directionality is top-down, or if we're inverting the direction
+        // because we're the LCA, commit a top-down message.
+        if (applyType == IPCMsgType.TopDown || isLCA) {
+            ++s.appliedTopDownNonce;
```

Comment:
Alfonso pointed out that this might be wrong. We'll need to revisit this:

https://filecoinproject.slack.com/archives/C05J93G9G4F/p1705391013362889